### PR TITLE
Proxy

### PR DESCRIPTION
**Pull request by mattcollier**

### CODE REVIEW

File: test/mocha/10-api.js [hunk 1]
```diff
@@ -0,0 +1,14 @@
+/*!
+ * Copyright (c) 2020 Digital Bazaar, Inc. All rights reserved.
+ */
+'use strict';
+
+const {agent, httpsAgent} = require('bedrock-https-agent');
```
Comment (dlongley) [inline test/mocha/10-api.js:6]: I believe this is only working because it runs after `bedrock.init`.

File: lib/index.js [hunk 2]
```diff
@@ -4,14 +4,20 @@
...
-const {config} = bedrock;
+const {config, util: {BedrockError}} = bedrock;
 const https = require('https');
 
 // module API
 const api = {};
-module.exports = api;
 
-api.httpsAgent = api.agent = null;
+module.exports = new Proxy(api, {
```
Comment (dlongley) [inline lib/index.js]: I don't think this is right. I think we want to wrap `httpsAgent` instead ... I'll look around for a pattern.

Comment (dlongley) [reply lib/index.js]: I think we need something more like this:

https://gist.github.com/dlongley/b34ab693e531dc892c9ac3ebed1cc6b7

Comment (mattcollier) [reply lib/index.js]: How is this different better?

Comment (dlongley) [reply lib/index.js]: What we have currently in this PR doesn't actually work. And the tests don't test it properly.

File: test/mocha/10-api.js (hunk 1 above)
Comment (dlongley) [reply test/mocha/10-api.js:6]: We may have to do something like this: https://github.com/digitalbazaar/bedrock-tokenization/blob/master/test/mocha/helpers.js#L52-L55

To test this.

File: lib/index.js [hunk 3]
```diff
@@ -4,14 +4,43 @@
...
+const {config, util: {BedrockError}} = bedrock;
 const https = require('https');
 
 // module API
 const api = {};
 module.exports = api;
 
-api.httpsAgent = api.agent = null;
+let _httpsAgent = null;
+const _error = new BedrockError('The agent is not ready.', 'InvalidStateError');
```
Comment (dlongley) [inline lib/index.js]:
```suggestion
const _error = () => new BedrockError('The agent is not ready.', 'InvalidStateError');
```

File: lib/index.js [hunk 4]
```diff
@@ -4,14 +4,43 @@
...
+  },
+  getOwnPropertyDescriptor(target, prop) {
+    if(!_httpsAgent) {
+      throw _error;
+    }
+    return Object.getOwnPropertyDescriptor(_httpsAgent, prop);
+  },
+  set(target, prop, value) {
+    if(!_httpsAgent) {
+      throw _error;
```
Comment (dlongley) [inline lib/index.js]:
```suggestion
    if(!_httpsAgent) {
      throw _error();
    }
    return _httpsAgent[propKey];
  },
  getPrototypeOf() {
    if(!_httpsAgent) {
      throw _error();
    }
    return Object.getPrototypeOf(_httpsAgent);
  },
  getOwnPropertyDescriptor(target, prop) {
    if(!_httpsAgent) {
      throw _error();
    }
    return Object.getOwnPropertyDescriptor(_httpsAgent, prop);
  },
  set(target, prop, value) {
    if(!_httpsAgent) {
      throw _error();
```

**Review by dlongley — APPROVED** [review]
Thanks!

**Pull request merged**